docs/docs/contributors/FSharp.md

```diff
@@ -0,0 +1,128 @@
+- [F# for Fun and Profit](https://fsharpforfunandprofit.com/)
+- [F# Fundamentals Tutorial | Learn Functional Programming | Step-by-Step Guide](https://www.youtube.com/watch?v=SvOInBxPL30)
+- [F# Foundation Slack](https://fsharp.org/guides/slack/)
+- [F# on Discord](https://discord.com/invite/R6n7c54)
+
+## Used F# features
+
+F# has a lot of nice language features, although not all of them are used in Fantomas.
+We wish to highlight the most important ones that we use before continuing:
+
+- [Partial active pattern](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns#partial-active-patterns), these heavily used in `SourceParser.fs`.
+  In short, we use the `Untyped Abstract Syntax Tree` created by the F# parser, we don't use all the information in that tree to restore the source code.
+
+For example [SynExpr.For](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synexpr.html#For), the definition looks like:
+
+```fsharp
+type SynExpr =
+    ...
+    /// F# syntax: 'for i = ... to ... do ...'
+    | For of
+        forDebugPoint: DebugPointAtFor *
+        toDebugPoint: DebugPointAtInOrTo *
+        ident: Ident *
+        equalsRange: range option *
+        identBody: SynExpr *
+        direction: bool *
+        toBody: SynExpr *
+        doBody: SynExpr *
+        range: range
+```
+
+  However, in Fantomas we have a partial active pattern that we use to easily grab the information we need from the AST.
+```fsharp
+let (|For|_|) =
+    function
+    | SynExpr.For (_, _, ident, equalsRange, e1, isUp, e2, e3, _) -> Some(ident, equalsRange, e1, e2, e3, isUp)
+    | _ -> None  
+```
+
+Notice the underscores, we don't use the `DebugPoint` info and `range`, so we drop that information in the result of the partial active pattern.
+We then later use this pattern to work with the AST we need.
+Example usage in `CodePrinter.fs`:
+
+```fsharp
+match expr with
+| For (ident, equalsRange, e1, e2, e3, isUp) ->
+    atCurrentColumn (
+        !- "for "
+        +> genIdent ident
+        +> genEq SynExpr_For_Equals equalsRange
+        +> sepSpace
+        +> genExpr astContext e1
+        +> ifElse isUp (!- " to ") (!- " downto ")
+        +> genExpr astContext e2
+        +> !- " do"
+        +> indent
+        +> sepNln
+        +> genExpr astContext e3
+        +> unindent
+    )
+| _ -> ...
+```
+
+- [Custom operators](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/operator-overloading#creating-new-operators). In F# there are some special operators like [|>](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-operators.html#(|%3E)) and [>>](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-operators.html#(%3E%3E)).  
+
+Note that these are just function themselves as well. Instead of specifying all the arguments after the function name, (infix) operators let you specify an argument before the operator and after. 
```

Comment:
```suggestion
Note that these are just functions themselves as well. Instead of specifying all the arguments after the function name, (infix) operators let you specify an argument before the operator and after. 
```